Unexpected New (...) reached GenBCode

## Compiler version

3.2.1

## Minimized code

macros.scala

```scala
//> using scala "3.2.1"
import scala.quoted._
import scala.annotation.experimental

object macros {

  trait Suspend[A[_[_]]] {
    def sus[F[_]](fa: F[A[F]]): A[F]
  }

  inline def mkSuspend[Alg[_[_]]]: Suspend[Alg] = ${ mkSuspendImpl[Alg] }

  @experimental
  def mkSuspendImpl[Alg[_[_]]: Type](using Quotes): Expr[Suspend[Alg]] = {
    import quotes.reflect._
    def instance[F[_]: Type](fExpr: Expr[F[Alg[F]]]): Expr[Alg[F]] = {
      val cl = Symbol.newClass(
        Symbol.spliceOwner,
        "$anon",
        parents = List(
          TypeRepr.of[Object],
          TypeRepr.of[Alg].appliedTo(TypeRepr.of[F]),
        ),
        decls = _ => Nil,
        selfType = None,
      )

      val tree = New(
        TypeTree.of[Alg]
      )

      Block(
        List(
          ClassDef(cl, Nil, Nil)
        ),
        New(
          TypeTree.ref(cl)
        ),
      ).asExprOf[Alg[F]]
    }

    '{
      new Suspend[Alg] {
        def sus[F[_]](fa: F[Alg[F]]): Alg[F] = ${ instance('fa) }
      }
    }
  }

}
```

main.scala

```scala
//> using scala "3.2.1"
import macros.Suspend

object demo extends App {

  trait MyAlg[F[_]] {
    def hello(s: String): F[Int] = ???
  }

  macros.mkSuspend[MyAlg]
}
```

## Output (click arrow to expand)
<details>

```scala
Error: Unexpected New(... with ... {...}/TypeTree[TypeRef(NoPrefix,class demo$$anon$2)]) reached GenBCode.
  Call was genLoad(New(TypeTree[TypeRef(NoPrefix,class demo$$anon$2)]),Ldemo$MyAlg;)
Error compiling project (Scala 3.2.1, JVM)
Error: Unexpected error when compiling project_f0e8bc45bf: 'Unexpected New(... with ... {...}/TypeTree[TypeRef(NoPrefix,class demo$$anon$2)]) reached GenBCode.
  Call was genLoad(New(TypeTree[TypeRef(NoPrefix,class demo$$anon$2)]),Ldemo$MyAlg;)'
Compilation failed
```
</details>

By compiling this code with `-Ycheck:all` we get the following error:

```scala
checking t/Test_2.scala after phase inlining
java.lang.AssertionError: assertion failed: `New` node must be wrapped in a `Select`:
  parent = new Object with demo.MyAlg[F] {...}:demo.MyAlg[F]
  child = new Object with demo.MyAlg[F] {...} while running Ycheck on t/Test_2.scala
exception occurred while compiling t/Test_2.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: `New` node must be wrapped in a `Select`:
  parent = new Object with demo.MyAlg[F] {...}:demo.MyAlg[F]
  child = new Object with demo.MyAlg[F] {...} while compiling t/Test_2.scala
java.lang.AssertionError: assertion failed: `New` node must be wrapped in a `Select`:
  parent = new Object with demo.MyAlg[F] {...}:demo.MyAlg[F]
  child = new Object with demo.MyAlg[F] {...}
```

--- a/compiler/src/dotty/tools/dotc/transform/TreeChecker.scala
+++ b/compiler/src/dotty/tools/dotc/transform/TreeChecker.scala
@@ -141,7 +141,7 @@ class TreeChecker extends Phase with SymTransformer {
       override def apply(parent: Tree, tree: Tree)(using Context): Tree = {
         tree match {
           case tree: New if !parent.isInstanceOf[tpd.Select] =>
-            assert(assertion = false, i"`New` node must be wrapped in a `Select`:\n  parent = ${parent.show}\n  child = ${tree.show}")
+            assert(assertion = false, i"`New` node must be wrapped in a `Select` of the constructor:\n  parent = ${parent.show}\n  child = ${tree.show}")
           case _: Annotated =>
             // Don't check inside annotations, since they're allowed to contain
             // somewhat invalid trees.

